torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -68,70 +68,106 @@ def __len__(self) -> int:
+        self.source_datapipe = source_datapipe
+        self.filepath_fn = filepath_fn
+        self.extra_check_fn = extra_check_fn
+        self._end_caching_flag: bool = False
+
+    def __iter__(self):
+        if self._end_caching_flag:
+            for d in self.source_datapipe:
+                yield d
+        else:
+            # In case of BC breaking, use RuntimeError for now. Warning is another option
+            raise RuntimeError("Please call `end_caching()` before iteration.")
+
+    def __add__(self, other_datapipe):
+        raise RuntimeError("`OnDiskCacheHolder` doesn't support add operation")
+
+    def _cache_check_fn(self, data):
+        filepath = self.filepath_fn(data)
+        if not os.path.exists(filepath):
+            return False
+        if self.extra_check_fn:
+            return self.extra_check_fn(filepath)
+        return True
+
+    def _end_caching(self):
+        todo_dp, cached_dp = self.source_datapipe.demux(2, self._cache_check_fn)
+        self._end_caching_flag = True
+        # Re-assign source_datapipe
+        self.source_datapipe = todo_dp
 
         # Cached: keeps filepath
-        cached_dp = cached_dp.map(fn=lambda data: data[0])
-
-        # Non-cached
-        # opDataPipe yield bytes
-        if not op_map:
-            todo_dp = opDataPipe(todo_dp.map(fn=lambda data: data[1]), *op_args, **op_kwargs).save_to_disk(
-                mode=mode, filepath_fn=filepath_fn
-            )
-        else:
-            todo_dp = (
-                opDataPipe(todo_dp.map(fn=lambda data: data[1]), *op_args, **op_kwargs)
-                .map(op_map)
-                .save_to_disk(mode=mode, filepath_fn=filepath_fn)
-            )
-
-        # Load files from local disk
-        return FileLoader(cached_dp.concat(todo_dp))
+        return cached_dp.map(fn=self.filepath_fn)
+
+
+@functional_datapipe("end_caching")
+class EndOnDiskCacheHolderIterDataPipe(IterDataPipe):
+    """
+    `EndOnDiskCacheHolder` is a IterDataPipe that indicates when theresult of
+    prior DataPipe will be saved local files specified by `filepath_fn` of
+    the corresponding `OnDiskCacheHolder`. And, the result is required to be
+    a tuple of file path and data.
+
+    Args:
+        mode: Mode in which cached files are opened for write the data. Binary mode by default.
+    """
+    def __new__(cls, datapipe, mode="wb"):
+        graph = traverse(datapipe, exclude_primitive=True)
+        cache_holder = EndOnDiskCacheHolderIterDataPipe._recursive_search(graph)
+        if cache_holder is None:
+            raise RuntimeError("Incomplete `OnDiskCacheHolder` is required in the pipeline before calling `end_caching` or `EndOnDiskCacheHolder`")
+        cached_dp = cache_holder._end_caching()
+        todo_dp = datapipe.save_to_disk(mode=mode, filepath_fn=cache_holder.filepath_fn)
+        return cached_dp.concat(todo_dp)
+
+    @staticmethod
+    def _recursive_search(graph):
+        for dp in graph.keys():
```

Comment:
Make sure you capture the closest one, in case if there are multiple things like:  url.on_disk_cache...end_caching...on_disk_cache..end_caching....